File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        
        def bfs(root):
            if not root:
                return []
            visited, ans, queue, count = set(), [], deque(), 0
            
            queue.append(root)
            ans.append([queue[0].val])
            
            while queue:
                n, inner = len(queue), []
                count += 1

                for i in range(n):
                    cur = queue.popleft()
                    if cur.left:
                        queue.append(cur.left)
                        inner.append(cur.left.val)
                    if cur.right:
                        queue.append(cur.right)
                        inner.append(cur.right.val)
                if inner and count % 2 == 0:
                    ans.append(inner)
                elif inner:
                    ans.append(inner[::-1])
                print('ans', ans)
                    
            return ans
        
        return bfs(root)

File: 103-binary-tree-zigzag-level-order-traversal/103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        bool rev = false;
        vector<vector<int>>ans;
        queue<TreeNode*>q;
        if (!root) return {};
        q.push(root);
        while (!q.empty()) {
            int n = q.size();
            vector<int> row;
            for (int i = 0; i < n; i++) {
                auto u = q.front();
                q.pop();
                row.push_back(u->val);
                if (u->left)
                    q.push(u->left);
                if (u->right)
                    q.push(u->right);
            }
            if (rev)
                reverse(row.begin(),row.end());
            rev = !rev;
            ans.push_back(row);
        }
        return ans;
    }
};